src/main/kotlin/com/atlassian/performance/tools/awsinfrastructure/api/AwsDatasetModification.kt

```diff
@@ -1,77 +1,59 @@
-package com.atlassian.performance.tools.awsinfrastructure
+package com.atlassian.performance.tools.awsinfrastructure.api
 
 import com.atlassian.performance.tools.aws.api.Aws
 import com.atlassian.performance.tools.aws.api.Investment
 import com.atlassian.performance.tools.aws.api.Resource
-import com.atlassian.performance.tools.awsinfrastructure.api.CustomDatasetSource
-import com.atlassian.performance.tools.awsinfrastructure.api.Infrastructure
-import com.atlassian.performance.tools.awsinfrastructure.api.InfrastructureFormula
-import com.atlassian.performance.tools.awsinfrastructure.api.ProvisionedInfrastructure
 import com.atlassian.performance.tools.awsinfrastructure.api.hardware.C5NineExtraLargeEphemeral
 import com.atlassian.performance.tools.awsinfrastructure.api.jira.Jira
 import com.atlassian.performance.tools.awsinfrastructure.api.jira.StandaloneFormula
 import com.atlassian.performance.tools.awsinfrastructure.api.virtualusers.AbsentVirtualUsersFormula
 import com.atlassian.performance.tools.infrastructure.api.dataset.Dataset
 import com.atlassian.performance.tools.infrastructure.api.distribution.PublicJiraSoftwareDistribution
+import com.atlassian.performance.tools.workspace.api.RootWorkspace
 import com.atlassian.performance.tools.workspace.api.TestWorkspace
 import org.apache.logging.log4j.LogManager.getLogger
 import java.time.Duration
+import java.util.*
 import java.util.concurrent.TimeUnit
+import java.util.function.Consumer
+import java.util.function.Function
 
-internal class AwsDataset(
-    private val dataset: Dataset
-) {
+class AwsDatasetModification private constructor(
```

Comment:
I do not see any cleanup code for extra plugins - is there decision to do it in transformer ?

Reply:
@dagguh has suggested to do cleaning inside the Consumer

Reply:
Ok,  should the consumer be aware of jira is still running while it needs to do some cleanup ?

Reply:
There should be hardcoded cleanup for license and plugins.
If the modification itself requires additional cleanup, then it has to be a part of the modification.

Reply:
Good call on the jira state. Some things might need cleanup after it is stopped. We can have `onlineModification` and `offlineModification`

Reply:
> Standard/default cleanup should be hardcoded in. It should include licenses and plugins.
How can we hardcode cleaning-up plugins if we don't know which plugins might have been provided via `InfrastructureFormula<*>`?

> We can have onlineModification and offlineModification
Should `offlineModification` be applied BEFORE Jira is started? Or should we stop Jira at some point and then apply this `offlineModification`?

Reply:
Nuke all the plugins.

TBH I don't know if we need pre-start or post-stop mods. Perhaps we're adding API unnecessarily? I don't know of any use-cases for those yet. And we're using the builder pattern, so it's trivial to add it in later when such use cases pop up.

Reply:
We don't know now - that's fair. But we may suspect or imagine such case. I think that replacing consumer with similar, but  homemade interface would provide us more flexibility in the future  
(we can simply add new method for some offline mod instead replacing interface)

Reply:
@drspartez current implementation doesn't block the flexibility in the future. There's no need to change the interface, we can add a new one.

Reply:
That is why i'd liek to avoid that. I do prefer keeping contract in one place. With multiple interfaces we make complexity. This is good practice (adding new interfaces for existing impl) to do change in legacy systems. Here we could look little bit far beyond. This cost nothing.

Reply:
You're right. It's a little change and gives us more room in the future.

Reply:
Add the `transformation` as the `AwsDatasetModification` field. Then the `Builder` allows us to add `preStartTransformation` and `postStopTransformation` in the future.
We also get rid of the `Function<>`

Reply:
OTOH if we predict dependencies between these transformations, then we might be tempted to bind them together in the `DatasetTransformation`. TBH I'm fine with both, but if we're planning to add methods to interfaces, then we gotta port it to Java.

Reply:
> Add the transformation as the AwsDatasetModification field. Then the Builder allows us to add preStartTransformation and postStopTransformation in the future.
We also get rid of the Function

That's exactly what I asked here:
https://github.com/atlassian/aws-infrastructure/pull/30#discussion_r298473750
And the conclusion was that "No-op seems not to be a good default"

Reply:
You can create a builder without default.

Reply:
The vessel for defaults (interface or builder) doesn't matter if there's no good default.
And if you want a new parameter/method, you have to either break compatibility or bring a default.

Reply:
I'm not sure if we are on the same page.
1. I'm ok with an interface. There are some problems with the interfaces/default methods and Java interop, but it's not a big deal. 
2. I'm ok with a builder. It will probably be more flexible, but we miss the one place we can inject a transformation that requires to be injected in multiple places of the infra's lifecycle.
3. I'm more or less ok with NoOp. It's not very bad, but the transformation with the default will be in most cases just a mistake. The builder can require a transformation, so it could eliminate a need for the default.

Reply:
ad. 1 Agreed
ad. 2 Agreed
ad. 3 Agreed. And if someone wants only to purge the dataset (e.g. me), they can easily pass in a NoOp themselves. We can provide a NoOp impl for them out-of-the-box to avoid copying the same boilerplate across clients.